### PR TITLE
[WIP][Reference] Refactor relationship creation inside CrudPanels 

### PR DESCRIPTION
**Pull request by pxpm**
Hey there!

This is a #3410 refactor.

The functionality remains the same as described in that PR thread. 

@tabacitu wanna give this a quick try and let me know your findings ? I tested this merging https://github.com/Laravel-Backpack/demo/tree/add-all-relationships into https://github.com/Laravel-Backpack/demo/tree/4.2

Heads-up for missing package to sanitize WYSIWYG inputs `mews/purifier`

Everything seems to be working accordingly, but I am a bit dizzy of looking/testing the same bit of code lots of times in the past days, maybe you can spot something that I can't!

I know how this is going to be, so shoot your questions, let your spidey sense flow and let's hope for the best 👍 

Pedro

### CODE REVIEW

**Comment by pxpm** [comment]
@tabacitu heads up, I've pushed `composer.json` here by error. What's the best way to remove it ? Just revert the changes in composer.json and make a new commit ?

File: composer.json [hunk 1]
```diff
@@ -37,7 +37,8 @@
         "creativeorange/gravatar": "~1.0",
         "composer/package-versions-deprecated": "^1.8",
         "doctrine/dbal": "^2.5|^3.0",
-        "guzzlehttp/guzzle": "^7.0|^6.3"
+        "guzzlehttp/guzzle": "^7.0|^6.3",
+        "mews/purifier": "^3.3"
```
Comment (tabacitu) [inline composer.json:41]:
```suggestion
        "guzzlehttp/guzzle": "^7.0|^6.3"
```

**Comment by tabacitu** [comment]
Hey Pedro,

I have good news and bad news 😅 👀 

Good news - I agree to fixing all of these.
Bad news - I think we've collaborated about this the wrong way 😔  and it'll take a little more work to have these merged;

I know you've worked very hard on this PR (and all previous ones regarding this), and I super-appreciate it. Nobody knows more than you about how this part of Backpack works. And I like it that way 😀  But as it is now, I'm sorry to say but... **this PR is close to impossible to review** 😅  

I've looked at this for hours, multiple times, both yesterday and today and... there's one big problem: **I can't match each change to its WHY.** As it is right now, I have no idea of the possible negative implications of merging this. So I'd be super-scared to merge this. It's not just a refactor, it's a refactor + a few features. VERY COMPLICATED features. MULTIPLE very complicated features. Explained in MANY issues, dating back years and years: 
- https://github.com/Laravel-Backpack/CRUD/pull/3410
- https://github.com/Laravel-Backpack/CRUD/pull/3272
- https://github.com/Laravel-Backpack/CRUD/pull/3196
- https://github.com/Laravel-Backpack/CRUD/pull/3250
- https://github.com/Laravel-Backpack/CRUD/pull/1453
- https://github.com/Laravel-Backpack/CRUD/pull/3196
- https://github.com/Laravel-Backpack/CRUD/issues/1442
- https://github.com/Laravel-Backpack/CRUD/issues/659#issuecomment-690036882
- https://github.com/Laravel-Backpack/CRUD/issues/2708

And to fix those unclear issues we need to make change in VERY IMPORTANT bits of Backpack logic. I've tried to write a round-up of the problems & solutions, but that proved to be impossible too. **I think you're the only one who can untangle this, because right now the "_big picture_" regarding these changes lives in one place and one place only - your head 😅**

I wanted to ask you to write a round-up, to draw me the big picture. But you've already done something similar to that [here](https://github.com/Laravel-Backpack/CRUD/pull/3410). And I think what that proves is that... even that is not easy to follow/understand/test/review. It needed this follow-up PR, after all. So here's what I think we can do.

**I think I made a mistake letting you work so much on one single PR, without the reviewer. We should have structured this into multiple small PRs to begin with**, so that it's not so scary to merge or difficult to test & review. But I've learned from this mistake 😀  I've just [added a PR template](https://github.com/Laravel-Backpack/CRUD/pull/3982) which I think will:
- help us test & review PRs a lot faster;
- force us into creating smaller PRs; as I said in that thread - we have a new rule of thumb: **If it's too difficult to _explain_ the PR... it's probably to difficult to _review_ the PR too.**

Please forgive me for asking you to work on this some more, just for my sake. But this is a BIG change and I think we _absolutely need_ confidence when merging all these changes. Please help me gain that confidence. 

**Here's what I propose we do:**
- [ ] Step 1. 📄    Please create a list of problems this PR fixes (_X relationship doesn't work, Y relationship behaves z not t_).
- [ ] Step 2. ⛏   Create one small PR, that fixes one problem (ideally the smallest one).
- [ ] Step 3. ✍️   Carefully write out the PR details using the new PR template.
- [ ] Step 4. ⚖️   Wait. We talk about that PR, review it, merge it into 4.2. 
- [ ] Step 5. 🔙   Go back to step 2 and tackle the next problem in the list (PR then review then merge).

**Sounds like a good plan?**

In the meantime, let's keep this PR open. After each PR gets merged, this PR should get smaller and smaller, until there's nothing there (or only contains changes that would make it worse).

What do you say @pxpm ? **Can we do this or can we _do this_ 💪 ?** Sorry to make you work extra-hard for this, but it's a big one and I think merging this as it is now will create problems for future-us.

Cheers!

**Comment by pxpm** [comment]
Hey @tabacitu 

Nothing I was not expecting, I know it looks like alot, but it scares more than it really is. 

I will rigth way start splitting this into smaller PR's.

Best,
Pedro

**Comment by tabacitu** [comment]
Thank you for being so understanding, Pedro 🙏  
Just one PR please. One. After that's merged into 4.2, we'll create another one.

I'm trying to avoid ending up with 4-5 different but interdependent PRs - that's wouldn't be good either.

**Comment by pxpm** [comment]
@tabacitu Ok, that one is done, please take a look at #3981 I was about to start from that branch as I think that is something we were should merge. That one will clean-up this one alot with all naming changes beeing removed.

**Comment by pxpm** [comment]
I think we can safely close this branch now so ... here goes BOOM!

**Pull request closed without merging**